### PR TITLE
Driver: Implement module handling in a helper

### PR DESCRIPTION
**Pull request by jakelang**

### CODE REVIEW

File: chisel/src/driver.rs [hunk 1]
```diff
@@ -184,6 +205,278 @@ impl ChiselDriver {
...
+        &mut self,
+        name: String,
+        module: ModuleConfig,
+        wasm: &mut Module,
+    ) -> Result<ModuleResult, DriverError> {
+        let result = match name.as_str() {
+            "checkfloat" => {
+                let checkfloat = CheckFloat::new();
+                let module_result = checkfloat.validate(wasm);
+                ModuleResult::Validator(name.clone(), module_result)
```
Comment (axic) [inline chisel/src/driver.rs]: Why not wrap these in `Ok()` and remove the match assignment?

File: chisel/src/driver.rs [hunk 2]
```diff
@@ -184,6 +205,278 @@ impl ChiselDriver {
...
+                    ModuleResult::Validator(name.clone(), module_result)
+                } else {
+                    chisel_debug!(1, "checkstartfunc missing field 'require_start'");
+                    return Err(DriverError::MissingRequiredField(
+                        name.clone(),
+                        "require_start".to_string(),
+                    ));
+                }
+            }
+            "deployer" => unimplemented!("Creator modules are not supported yet."),
```
Comment (axic) [inline chisel/src/driver.rs]: This is not a creator anymore.

File: chisel/src/driver.rs [hunk 3]
```diff
@@ -184,6 +205,278 @@ impl ChiselDriver {
...
+            "dropsection" => {
```
Comment (axic) [inline chisel/src/driver.rs]: Suggest this implementation is dropped as-is given its not there in the current CLI.

File: chisel/src/driver.rs [hunk 4]
```diff
@@ -184,6 +205,278 @@ impl ChiselDriver {
...
+                    (None, _) => {
+                        chisel_debug!(1, "dropsection missing field 'mode'");
+                        return Err(DriverError::MissingRequiredField(
+                            name.clone(),
+                            "mode".to_string(),
+                        ));
+                    }
+                }
+            }
+            "fromwat" => unimplemented!("Creator modules are not supported yet."),
```
Comment (axic) [inline chisel/src/driver.rs]: Can be dropped for same reason.

File: chisel/src/driver.rs [hunk 5]
```diff
@@ -184,6 +205,278 @@ impl ChiselDriver {
...
+                        ));
+                    }
+                } else {
+                    chisel_debug!(1, "verifyimports missing field 'preset'");
+                    return Err(DriverError::MissingRequiredField(
+                        name.clone(),
+                        "preset".to_string(),
+                    ));
+                }
+            }
```
Comment (axic) [inline chisel/src/driver.rs]: Please add `dropnames` and `binaryenopt` though for the moment.

File: chisel/src/driver.rs [hunk 6]
```diff
@@ -184,6 +205,278 @@ impl ChiselDriver {
         self.state = DriverState::Done(results);
         &self.state
     }
+
+    pub fn execute_module(
+        &mut self,
+        name: String,
```
Comment (axic) [inline chisel/src/driver.rs]: Can't this be a `&str`?

Comment (axic) [reply chisel/src/driver.rs]: Ah ok, you're consuming `name` and `module` from the caller.

File: chisel/src/driver.rs (hunk 1 above)
Comment (axic) [inline chisel/src/driver.rs]: Can't you just consume `name` here (and the rest of the places below)?

Comment (jakelang) [reply chisel/src/driver.rs]: It'd be more concise but I think it looks a bit strange to have an `Ok()` wrapping an expression which could actually short-circuit and return `Err()` mid-evaluation

File: chisel/src/driver.rs (hunk 5 above)
Comment (jakelang) [reply chisel/src/driver.rs]: We have a module called `dropnames`?

Comment (axic) [reply chisel/src/driver.rs]: No, but we have a case in the current CLI called `dropnames`.

File: chisel/src/driver.rs [hunk 7]
```diff
@@ -184,6 +195,235 @@ impl ChiselDriver {
...
+                } else {
+                    chisel_debug!(1, "binaryenopt missing field 'preset'");
+                    return Err(DriverError::MissingRequiredField(
+                        name,
+                        "preset".to_string(),
+                    ));
+                }
+            }
+            other => {
+                return Err(DriverError::ModuleNotFound(name.clone(), other.to_string()));
```
Comment (axic) [inline chisel/src/driver.rs]: No need for clone here?

File: chisel/src/driver.rs [hunk 8]
```diff
@@ -184,6 +195,235 @@ impl ChiselDriver {
...
+                } else {
+                    chisel_debug!(1, "verifyimports missing field 'preset'");
+                    return Err(DriverError::MissingRequiredField(
+                        name,
+                        "preset".to_string(),
+                    ));
+                }
+            }
+            #[cfg(feature = "binaryen")]
+            "binaryenopt" => {
```
Comment (axic) [inline chisel/src/driver.rs]: Doesn't make any difference but this it out of lexicographic order (while the rest is in order).

File: chisel/src/driver.rs (hunk 7 above)
Comment (jakelang) [reply chisel/src/driver.rs]: It is needed, since it's in a scope where name is borrowed. See line 206

File: chisel/src/driver.rs (hunk 8 above)
Comment (jakelang) [reply chisel/src/driver.rs]: I thought it would make more sense to separate conditionally compiled cases.

File: chisel/src/driver.rs (hunk 7 above)
Comment (jakelang) [reply chisel/src/driver.rs]: NVM, it is actually needed because we use `other.to_string()` as the second argument.

**Pull request merged**